optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/change/DefaultProblemChangeDirector.java

```diff
@@ -0,0 +1,88 @@
+import org.optaplanner.core.api.solver.change.ProblemChangeDirector;
+
+public final class DefaultProblemChangeDirector<Solution_> implements ProblemChangeDirector {
+
+    private final ScoreDirector<Solution_> scoreDirector;
+
+    public DefaultProblemChangeDirector(ScoreDirector<Solution_> scoreDirector) {
+        this.scoreDirector = scoreDirector;
+    }
+
+    @Override
+    public <Entity> void addEntity(Entity entity, Consumer<Entity> entityConsumer) {
+        scoreDirector.beforeEntityAdded(entity);
+        entityConsumer.accept(entity);
+        scoreDirector.afterEntityAdded(entity);
+    }
+
+    @Override
+    public <Entity> void removeEntity(Entity entity, Consumer<Entity> workingEntityConsumer) {
+        Entity workingEntity = lookUpWorkingObject(entity);
+        scoreDirector.beforeEntityRemoved(workingEntity);
+        workingEntityConsumer.accept(workingEntity);
+        scoreDirector.afterEntityRemoved(workingEntity);
+    }
+
+    @Override
+    public <Entity> void changeVariable(Entity entity, Consumer<Entity> workingEntityConsumer, String variableName) {
+        Entity workingEntity = lookUpWorkingObject(entity);
+        scoreDirector.beforeVariableChanged(workingEntity, variableName);
+        workingEntityConsumer.accept(workingEntity);
+        scoreDirector.afterVariableChanged(workingEntity, variableName);
+    }
+
+    @Override
+    public <ProblemFact> void addProblemFact(ProblemFact problemFact, Consumer<ProblemFact> problemFactConsumer) {
+        scoreDirector.beforeProblemFactAdded(problemFact);
+        problemFactConsumer.accept(problemFact);
+        scoreDirector.afterProblemFactAdded(problemFact);
+    }
+
+    @Override
+    public <ProblemFact> void removeProblemFact(ProblemFact problemFact, Consumer<ProblemFact> workingProblemFactConsumer) {
+        ProblemFact workingProblemFact = lookUpWorkingObject(problemFact);
+        scoreDirector.beforeProblemFactRemoved(workingProblemFact);
+        workingProblemFactConsumer.accept(workingProblemFact);
+        scoreDirector.afterProblemFactRemoved(workingProblemFact);
+    }
+
+    @Override
+    public <EntityOrProblemFact> void changeProblemProperty(EntityOrProblemFact problemFactOrEntity,
+            Consumer<EntityOrProblemFact> workingProblemFactOrEntityConsumer) {
+        EntityOrProblemFact workingEntityOrProblemFact = lookUpWorkingObject(problemFactOrEntity);
+        scoreDirector.beforeProblemPropertyChanged(workingEntityOrProblemFact);
+        workingProblemFactOrEntityConsumer.accept(workingEntityOrProblemFact);
+        scoreDirector.afterProblemPropertyChanged(workingEntityOrProblemFact);
+    }
+
+    @Override
+    public <E> E lookUpWorkingObject(E externalObject) {
+        return scoreDirector.lookUpWorkingObject(externalObject);
+    }
+
+    @Override
+    public <E> E lookUpWorkingObjectOrReturnNull(E externalObject) {
+        return scoreDirector.lookUpWorkingObjectOrReturnNull(externalObject);
+    }
```

Comment:
Great use case for `Optional`. No need to have two methods, let the user choose the behavior they want.

Reply:
Good point, we can change the existing methods now.

Reply:
At a second glance, this might not be the best approach:
1) it requires an extra action on the user's side
2) more importantly, we lose a precise error message from the `LookUpStrategy` implementation (due to calling its method that does not assert the lookup's result)